File: backend/services/routerService.js
const routerConfig = require('../config/router');
const fs = require('fs').promises;
const path = require('path');
const { saveSession, getSessionByToken, getSessionByPhone, expireSession, getAllActive } = require('../db');

class RouterService {
  constructor() {
    this.routerType = process.env.ROUTER_TYPE || 'mikrotik';
    this.config = routerConfig[this.routerType];
    // Keep a lightweight in-memory index for fast lookups (DB is source of truth)
    this.activeConnections = new Map();
    this._loadFromDB();
  }

  // Restore active sessions from DB on startup
  _loadFromDB() {
    try {
      const rows = getAllActive.all();
      rows.forEach(row => {
        this.activeConnections.set(row.token, {
          token: row.token,
          phoneNumber: row.phone,
          packageName: row.package,
          packagePrice: row.price,
          userIP: row.user_ip,
          macAddress: row.mac_address,
          startTime: row.start_time,
          expiryTime: row.expiry_time,
          durationMinutes: row.duration,
          transactionId: row.txn_id,
        });
      });
      if (rows.length > 0) console.log(`📦 Restored ${rows.length} active session(s) from database`);
    } catch (e) {
      console.warn('Could not load sessions from DB:', e.message);
    }
  }

  async activateConnection(data) {
    switch (this.routerType) {
      case 'mikrotik': return await this.activateMikroTik(data);
      case 'openwrt':  return await this.activateOpenWrt(data);
      case 'pfsense':  return await this.activatePfSense(data);
      default:         return await this.activateGeneric(data);
    }
  }

  // ─── MikroTik ────────────────────────────────────────────────────────────────
  async activateMikroTik(data) {
    try {
      console.log(`📡 MikroTik: connecting to ${this.config.host}`);

      const simulateConnection = process.env.NODE_ENV === 'development' && !this.config.password;
      if (simulateConnection) {
        console.log(`⚠️  Simulating — would whitelist MAC:${data.macAddress || 'unknown'} IP:${data.userIP} for ${data.durationMinutes}m`);
        this.storeConnection(data, `mikrotik_${data.transactionId}`);
        return { success: true, token: `mikrotik_${data.transactionId}`, sessionId: `session_${Date.now()}`, message: 'Simulated' };
      }

      let RouterOSAPI;
      try { RouterOSAPI = require('routeros-api'); }
      catch (e) {
        this.storeConnection(data, `mikrotik_${data.transactionId}`);
        return { success: true, token: `mikrotik_${data.transactionId}`, sessionId: `session_${Date.now()}`, message: 'routeros-api not installed' };
      }

      const { RouterOSClient } = RouterOSAPI;
      const api = new RouterOSClient({
        host: this.config.host, user: this.config.username,
        password: this.config.password, port: this.config.port || 8728, timeout: 10
      });

      try {
        const client = await api.connect();
        const timeout = data.durationMinutes || 60;
        const comment = `GORNHOM|${data.phoneNumber}|${data.packageName}|${data.transactionId}`;

        // Prefer MAC address — survives IP changes on reconnect
        const identifier = data.macAddress || data.userIP;

        const addOrUpdate = async (list, address) => {
          try {
            await client.menu('/ip firewall address-list').add({ list, address, timeout: `${timeout}m`, comment });
          } catch (e) {
            if (e.message && e.message.includes('already have')) {
              const entries = await client.menu('/ip firewall address-list')
                .where('list', list).where('address', address).get();
              if (entries.length > 0) {
                await client.menu('/ip firewall address-list').where('id', entries[0].id).update({ timeout: `${timeout}m`, comment });
              }
            } else throw e;
          }
        };

        // Whitelist both MAC (for reconnect) and IP (for immediate access)
        if (data.macAddress) await addOrUpdate('allowed-macs', data.macAddress);
        await addOrUpdate('allowed-users', data.userIP);

        api.close();
        this.storeConnection(data, `mikrotik_${data.transactionId}`);
        return { success: true, token: `mikrotik_${data.transactionId}`, sessionId: `session_${Date.now()}` };
      } catch (e) { api.close(); throw e; }
    } catch (error) {
      console.error('❌ MikroTik error:', error.message);
      return { success: false, message: error.message };
    }
  }

  // ─── OpenWrt ─────────────────────────────────────────────────────────────────
  async activateOpenWrt(data) {
    try {
      const { exec } = require('child_process');
      const { promisify } = require('util');
      const execAsync = promisify(exec);

      // Use MAC-based rule if available, fall back to IP
      let rule;
      if (data.macAddress) {
        rule = `iptables -I FORWARD -m mac --mac-source ${data.macAddress} -j ACCEPT`;
      } else {
        rule = `iptables -I FORWARD -s ${data.userIP} -j ACCEPT && iptables -I FORWARD -d ${data.userIP} -j ACCEPT`;
      }

      await execAsync(`ssh -o StrictHostKeyChecking=no -o ConnectTimeout=5 ${this.config.username}@${this.config.host} "${rule}"`);
      this.storeConnection(data, `openwrt_${data.transactionId}`);
      return { success: true, token: `openwrt_${data.transactionId}`, sessionId: `session_${Date.now()}` };
    } catch (error) {
      return { success: false, message: `SSH failed: ${error.message}` };
    }
  }

  // ─── pfSense ─────────────────────────────────────────────────────────────────
  async activatePfSense(data) {
    try {
      const axios = require('axios');
      const response = await axios.post(
        `https://${this.config.host}/api/v1/firewall/rule`,
        { interface: 'lan', type: 'pass', source: data.userIP, destination: 'any',
          description: `GORNHOM|${data.phoneNumber}|${data.packageName}` },
        { auth: { username: this.config.username, password: this.config.password }, timeout: 5000 }
      );
      if (!response.data.success) throw new Error(response.data.message);
      this.storeConnection(data, `pfsense_${data.transactionId}`);
      return { success: true, token: `pfsense_${data.transactionId}`, sessionId: `session_${Date.now()}` };
    } catch (error) {
      return { success: false, message: error.message };
    }
  }

  // ─── Generic (log only) ──────────────────────────────────────────────────────
  async activateGeneric(data) {
    try {
      const logDir = path.join(__dirname, '../logs');
      await fs.mkdir(logDir, { recursive: true });
      await fs.appendFile(path.join(logDir, 'connections.log'), JSON.stringify({ timestamp: new Date().toISOString(), action: 'ACTIVATE', ...data }) + '\n');
      this.storeConnection(data, `generic_${data.transactionId}`);
      return { success: true, token: `generic_${data.transactionId}`, sessionId: `session_${Date.now()}` };
    } catch (error) {
      return { success: false, message: error.message };
    }
  }

  // ─── Reconnect: find active session by phone, re-whitelist new IP/MAC ────────
  async reconnectByPhone(phoneNumber, newIP, newMAC) {
    // Check DB first (survives restarts)
    let conn = null;
    try {
      const row = getSessionByPhone.get(phoneNumber);
      if (row) {
        conn = {
          token: row.token,
          phoneNumber: row.phone,
          packageName: row.package,
          packagePrice: row.price,
          userIP: row.user_ip,
          macAddress: row.mac_address,
          expiryTime: row.expiry_time,
          durationMinutes: row.duration,
          transactionId: row.txn_id,
        };
      }
    } catch (e) { /* fall through to memory */ }

    // Fall back to in-memory
    if (!conn) {
      for (const c of this.activeConnections.values()) {
        if (c.phoneNumber === phoneNumber) { conn = c; break; }
      }
    }

    if (!conn) return { success: false, message: 'No active session for this phone number' };

    const remaining = new Date(conn.expiryTime).getTime() - Date.now();
    if (remaining <= 0) {
      expireSession.run(conn.token);
      this.activeConnections.delete(conn.token);
      return { success: false, message: 'Session has expired. Please purchase a new package.' };
    }

    const remainingMinutes = Math.ceil(remaining / 60000);
    console.log(`🔄 Reconnecting ${phoneNumber} — ${remainingMinutes}m remaining`);

    await this.activateConnection({
      ...conn,
      userIP: newIP || conn.userIP,
      macAddress: newMAC || conn.macAddress,
      durationMinutes: remainingMinutes,
      transactionId: conn.transactionId
    });

    return {
      success: true,
      connectionToken: conn.token,
      remainingMinutes,
      expiresAt: conn.expiryTime,
      packageName: conn.packageName,
      message: `Reconnected — ${remainingMinutes} minutes remaining`
    };
  }

  // ─── Store session ────────────────────────────────────────────────────────────
  storeConnection(data, token) {
    let maxDevices = parseInt(process.env.MAX_DEVICES_PER_SESSION) || 1;
    try {
      const { getPackages } = require('../routes/packages');
      const pkg = getPackages().find(p => p.name === data.packageName);
      if (pkg) maxDevices = pkg.maxDevices;
    } catch (e) { /* packages route not loaded yet, use default */ }

    const expiryTime = data.expiryTime || new Date(Date.now() + data.durationMinutes * 60 * 1000).toISOString();
    const startTime = new Date().toISOString();

    const connection = {
      token,
      userIP: data.userIP,
      macAddress: data.macAddress || null,
      phoneNumber: data.phoneNumber,
      packageName: data.packageName,
      packagePrice: data.packagePrice,
      startTime,
      expiryTime,
      durationMinutes: data.durationMinutes,
      transactionId: data.transactionId,
      maxDevices,
      connectedDevices: [data.macAddress || data.userIP].filter(Boolean)
    };

    // Persist to database
    try {
      saveSession.run({
        token,
        phone: data.phoneNumber,
        package: data.packageName,
        price: data.packagePrice || 0,
        currency: data.packageCurrency || 'KES',
        duration: data.durationMinutes,
        userIP: data.userIP,
        macAddress: data.macAddress || null,
        txnId: data.transactionId,
        paymentMethod: data.paymentMethod || 'paystack',
        startTime,
        expiryTime,
      });
    } catch (e) {
      console.warn('DB save failed, using memory only:', e.message);
    }

    this.activeConnections.set(token, connection);

    // Auto-expire from memory (DB handles persistence)
    const ttl = new Date(expiryTime).getTime() - Date.now();
    if (ttl > 0) {
      setTimeout(() => {
        this.activeConnections.delete(token);
        try { expireSession.run(token); } catch (e) {}
        console.log(`⏰ Session expired: ${token}`);
      }, ttl);
    }
  }

  // ─── Check if a new device can join an existing session ──────────────────────
  canAddDevice(token, deviceIdentifier) {
    const conn = this.activeConnections.get(token);
    if (!conn) return { allowed: false, reason: 'Session not found' };
    if (conn.connectedDevices.includes(deviceIdentifier)) return { allowed: true, reason: 'Already connected' };
    if (conn.connectedDevices.length >= conn.maxDevices) {
      return { allowed: false, reason: `Device limit reached (max ${conn.maxDevices})` };
    }
    conn.connectedDevices.push(deviceIdentifier);
    return { allowed: true, reason: 'Device added' };
  }

  async checkConnectionStatus(token) {
    // Check DB first
    let conn = this.activeConnections.get(token);
    if (!conn) {
      try {
        const row = getSessionByToken.get(token);
        if (row) conn = { expiryTime: row.expiry_time, packageName: row.package, userIP: row.user_ip, connectedDevices: [], maxDevices: 1 };
      } catch (e) {}
    }
    if (!conn) return { active: false, message: 'Session not found' };

    const remaining = new Date(conn.expiryTime).getTime() - Date.now();
    const remainingMinutes = Math.max(0, Math.floor(remaining / 60000));
    return {
      active: remainingMinutes > 0,
      remainingMinutes,
      expiresAt: conn.expiryTime,
      packageName: conn.packageName,
      userIP: conn.userIP,
      connectedDevices: conn.connectedDevices ? conn.connectedDevices.length : 1,
      maxDevices: conn.maxDevices || 1
    };
  }

  async revokeConnection(identifier) {
    for (const [token, conn] of this.activeConnections.entries()) {
      if (conn.userIP === identifier || token === identifier || conn.phoneNumber === identifier) {
        this.activeConnections.delete(token);
        try { expireSession.run(token); } catch (e) {}
        console.log(`🔒 Revoked session for ${conn.phoneNumber}`);
        return { success: true, message: 'Connection revoked', userIP: conn.userIP };
      }
    }
    return { success: false, message: 'Connection not found' };
  }
}

module.exports = new RouterService();

File: backend/services/router_service.go
package services

import (
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/gornhom/backend/config"
	"github.com/gornhom/backend/db"
	"github.com/gornhom/backend/models"
)

type ConnectionEntry struct {
	Token            string
	PhoneNumber      string
	PackageName      string
	PackagePrice     float64
	UserIP           string
	MacAddress       string
	StartTime        string
	ExpiryTime       string
	DurationMinutes  int
	TransactionID    string
	ConnectedDevices []string
	MaxDevices       int
}

// packageMaxDevices maps package names to their device limits
var packageMaxDevices = map[string]int{
	"30 Minutes":      1,
	"1 Hour":          1,
	"1 Day":           2,
	"1 Week":          3,
	"1 Month":         5,
	"Enterprise Plan": 25,
}

func maxDevicesForPackage(name string) int {
	if n, ok := packageMaxDevices[name]; ok {
		return n
	}
	return 1
}

type RouterService struct {
	cfg               *config.Config
	mu                sync.RWMutex
	activeConnections map[string]*ConnectionEntry // keyed by token
}

func NewRouterService(cfg *config.Config) *RouterService {
	rs := &RouterService{
		cfg:               cfg,
		activeConnections: make(map[string]*ConnectionEntry),
	}
	rs.loadFromDB()
	return rs
}

func (rs *RouterService) loadFromDB() {
	sessions, err := db.GetAllActive()
	if err != nil {
		log.Printf("Could not load sessions from DB: %v", err)
		return
	}
	for _, s := range sessions {
		entry := &ConnectionEntry{
			Token:           s.Token,
			PhoneNumber:     s.Phone,
			PackageName:     s.Package,
			PackagePrice:    s.Price,
			UserIP:          s.UserIP,
			MacAddress:      s.MacAddress,
			StartTime:       s.StartTime,
			ExpiryTime:      s.ExpiryTime,
			DurationMinutes: s.Duration,
			TransactionID:   s.TxnID,
			MaxDevices:      1,
		}
		rs.activeConnections[s.Token] = entry
		rs.scheduleExpiry(s.Token, s.ExpiryTime)
	}
	if len(sessions) > 0 {
		log.Printf("📦 Restored %d active session(s) from database", len(sessions))
	}
}

func (rs *RouterService) ActivateConnection(data *models.ActivationData) *models.ActivationResult {
	switch rs.cfg.RouterType {
	case "mikrotik":
		return rs.activateMikroTik(data)
	default:
		return rs.activateGeneric(data)
	}
}

func (rs *RouterService) activateMikroTik(data *models.ActivationData) *models.ActivationResult {
	token := fmt.Sprintf("mikrotik_%s", data.TransactionID)
	log.Printf("📡 MikroTik: would whitelist MAC:%s IP:%s for %dm", data.MacAddress, data.UserIP, data.DurationMinutes)
	// In production: use a MikroTik API library or SSH to add firewall address-list entries
	rs.storeConnection(data, token)
	return &models.ActivationResult{
		Success:   true,
		Token:     token,
		SessionID: fmt.Sprintf("session_%d", time.Now().UnixMilli()),
		Message:   "Connection activated",
	}
}

func (rs *RouterService) activateGeneric(data *models.ActivationData) *models.ActivationResult {
	token := fmt.Sprintf("generic_%s", data.TransactionID)
	log.Printf("📝 Generic: whitelist IP:%s MAC:%s for %dm", data.UserIP, data.MacAddress, data.DurationMinutes)
	rs.storeConnection(data, token)
	return &models.ActivationResult{
		Success:   true,
		Token:     token,
		SessionID: fmt.Sprintf("session_%d", time.Now().UnixMilli()),
		Message:   "Connection activated (generic)",
	}
}

func (rs *RouterService) storeConnection(data *models.ActivationData, token string) {
	now := time.Now().UTC()
	expiry := now.Add(time.Duration(data.DurationMinutes) * time.Minute)

	session := &models.Session{
		Token:         token,
		Phone:         data.PhoneNumber,
		Package:       data.PackageName,
		Price:         data.PackagePrice,
		Currency:      data.PackageCurrency,
		Duration:      data.DurationMinutes,
		UserIP:        data.UserIP,
		MacAddress:    data.MacAddress,
		TxnID:         data.TransactionID,
		PaymentMethod: data.PaymentMethod,
		StartTime:     now.Format(time.RFC3339),
		ExpiryTime:    expiry.Format(time.RFC3339),
	}
	if session.Currency == "" {
		session.Currency = "KES"
	}
	if session.PaymentMethod == "" {
		session.PaymentMethod = "paystack"
	}

	if err := db.SaveSession(session); err != nil {
		log.Printf("Failed to save session: %v", err)
	}

	entry := &ConnectionEntry{
		Token:           token,
		PhoneNumber:     data.PhoneNumber,
		PackageName:     data.PackageName,
		PackagePrice:    data.PackagePrice,
		UserIP:          data.UserIP,
		MacAddress:      data.MacAddress,
		StartTime:       session.StartTime,
		ExpiryTime:      session.ExpiryTime,
		DurationMinutes: data.DurationMinutes,
		TransactionID:   data.TransactionID,
		MaxDevices:      maxDevicesForPackage(data.PackageName),
	}

	rs.mu.Lock()
	rs.activeConnections[token] = entry
	rs.mu.Unlock()

	rs.scheduleExpiry(token, session.ExpiryTime)
}

func (rs *RouterService) scheduleExpiry(token, expiryTime string) {
	t, err := time.Parse(time.RFC3339, expiryTime)
	if err != nil {
		return
	}
	d := time.Until(t)
	if d <= 0 {
		return
	}
	time.AfterFunc(d, func() {
		rs.mu.Lock()
		delete(rs.activeConnections, token)
		rs.mu.Unlock()
		db.ExpireSession(token)
		log.Printf("⏰ Session expired: %s", token)
	})
}

func (rs *RouterService) CheckConnectionStatus(token string) map[string]interface{} {
	rs.mu.RLock()
	entry, ok := rs.activeConnections[token]
	rs.mu.RUnlock()

	if !ok {
		return map[string]interface{}{"active": false, "message": "Session not found"}
	}

	expiry, _ := time.Parse(time.RFC3339, entry.ExpiryTime)
	remaining := time.Until(expiry)
	if remaining <= 0 {
		return map[string]interface{}{"active": false, "message": "Session expired"}
	}

	return map[string]interface{}{
		"active":           true,
		"remainingMinutes": int(remaining.Minutes()),
		"expiresAt":        entry.ExpiryTime,
		"packageName":      entry.PackageName,
		"userIP":           entry.UserIP,
		"connectedDevices": len(entry.ConnectedDevices),
		"maxDevices":       entry.MaxDevices,
	}
}

func (rs *RouterService) RevokeConnection(identifier string) map[string]interface{} {
	rs.mu.Lock()
	defer rs.mu.Unlock()

	// Try by token first
	if _, ok := rs.activeConnections[identifier]; ok {
		delete(rs.activeConnections, identifier)
		db.ExpireSession(identifier)
		return map[string]interface{}{"success": true, "message": "Connection revoked"}
	}

	// Try by IP
	for token, entry := range rs.activeConnections {
		if entry.UserIP == identifier {
			delete(rs.activeConnections, token)
			db.ExpireSession(token)
			return map[string]interface{}{"success": true, "message": "Connection revoked", "userIP": identifier}
		}
	}

	return map[string]interface{}{"success": false, "message": "Connection not found"}
}

func (rs *RouterService) ReconnectByPhone(phone, newIP, newMAC string) map[string]interface{} {
	session, err := db.GetSessionByPhone(phone)
	if err != nil || session == nil {
		return map[string]interface{}{"success": false, "message": "No active session found for this phone number"}
	}

	expiry, _ := time.Parse(time.RFC3339, session.ExpiryTime)
	remaining := time.Until(expiry)
	if remaining <= 0 {
		return map[string]interface{}{"success": false, "message": "Session has expired"}
	}

	// Re-whitelist with new IP/MAC
	log.Printf("🔄 Reconnect: phone=%s newIP=%s newMAC=%s remaining=%.0fm", phone, newIP, newMAC, remaining.Minutes())

	return map[string]interface{}{
		"success":          true,
		"connectionToken":  session.Token,
		"remainingMinutes": int(remaining.Minutes()),
		"expiresAt":        session.ExpiryTime,
		"packageName":      session.Package,
		"message":          "Reconnected successfully",
	}
}

func (rs *RouterService) CanAddDevice(token, deviceID string) map[string]interface{} {
	rs.mu.RLock()
	entry, ok := rs.activeConnections[token]
	rs.mu.RUnlock()

	if !ok {
		return map[string]interface{}{"allowed": false, "reason": "Session not found"}
	}

	for _, d := range entry.ConnectedDevices {
		if d == deviceID {
			return map[string]interface{}{"allowed": true, "reason": "Device already in session"}
		}
	}

	if len(entry.ConnectedDevices) >= entry.MaxDevices {
		return map[string]interface{}{"allowed": false, "reason": "Device limit reached"}
	}

	rs.mu.Lock()
	entry.ConnectedDevices = append(entry.ConnectedDevices, deviceID)
	rs.mu.Unlock()

	return map[string]interface{}{"allowed": true, "reason": "Device added to session"}
}

func (rs *RouterService) GetActiveConnections() []*ConnectionEntry {
	rs.mu.RLock()
	defer rs.mu.RUnlock()
	entries := make([]*ConnectionEntry, 0, len(rs.activeConnections))
	for _, e := range rs.activeConnections {
		entries = append(entries, e)
	}
	return entries
}
